Bluetooth: Mesh: Shell: Add cfg opt for health srv

Andrewpini opened the pull request:
Adds Kconfig option for the Mesh Shell Health server.

The shell implementation of Health Server test commands is dependent on the user adding and initializing the shell Health Server instance in their application to make the commands functional. To prevent any confussion about how these commands work, a Kconfig option has been added to allow conditional compilation of these commands. In this way the user must actively enable this feature, making the conditional requirements more clear.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/mesh/shell/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# Mesh Shell configuration options
```
```suggestion
# Bluetooth mesh shell configuration options
```

PavelVPV reviewed (commented):
I don't have strong opinion on how this should be fixed. It would be good to hear others opinion. Maybe @omkar3141 could look at this?

Andrewpini commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -138,9 +139,14 @@ static const struct bt_mesh_health_srv_cb health_srv_cb = {
+
 
 struct bt_mesh_health_srv bt_mesh_shell_health_srv = {
+#if defined(CONFIG_BT_MESH_SHELL_HEALTH_SRV)
```
Isn't it? `health_srv_cb` won't be declared unless `CONFIG_BT_MESH_SHELL_HEALTH_SRV` is enabled, so I thought this was needed  to avoid compilation err.

Andrewpini commented:
> I don't have strong opinion on how this should be fixed. It would be good to hear others opinion. Maybe @omkar3141 could look at this?

Neither do I. It would be nice to provide something that is more dynamic, but I guess this is a "cheap" fix for now.

alxelax commented on subsys/bluetooth/mesh/shell/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# Bluetooth mesh shell configuration options
```
Is it worth creating a separate configuration file for one option?
What is the benefit of this idea?

Andrewpini replied:
For this option isolated I could agree to put it in the common Kconfig, but I highly suspect more shell options will be added in the future. In that case I would prefer to keep a separate file for it. I don't like the idea of dumping everything in a single file, since it IMO creates some noise for people trying to understand Mesh configuration. Shell is not a core feature of the mesh spec, so it might aswell keep it's configuration options in a separate file.

alxelax replied:
Which options are you going to move here?

alxelax replied:
Is it going to be separate PR ?

Andrewpini replied:
There are not currently any other option that needs to be moved. But if I start by adding one extra option to the main Kconfig now I suspect it will make a precedence where more options will be added there later.

alxelax replied:
IMHO this is a bit tricky. I'll not block PR due to this but I'd prefer changes for purpose.

PavelVPV replied:
I start thinking that we need a separate Kconfig. Because if we add more models to shell, this will clutter the main Kconfig. I also start thinking that we (will) need to add Kconfig options to control which mesh shell commands needs to be enabled, like we do in nRF Connect SDK https://github.com/nrfconnect/sdk-nrf/blob/main/subsys/bluetooth/mesh/shell/Kconfig. I think users should be able to choose which options they actually need.

alxelax commented on subsys/bluetooth/mesh/shell/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_MESH_SHELL_HEALTH_SRV
+	bool "Support for Shell Health Server Model"
```
Should it be default `y`?

Andrewpini replied:
IMO, no. Some of the point is to make the user aware that the health server features of shell requires some initialization. If the option is set by default the command will apear in the shell menu, even though it is useless due to lacking initialization.

omkar3141 commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -138,9 +139,14 @@ static const struct bt_mesh_health_srv_cb health_srv_cb = {
+
+
 
 struct bt_mesh_health_srv bt_mesh_shell_health_srv = {
```
I'm bit confused because of the description of the PR. Shouldn't the 'shell.c' automatically pick up the Health Server instance from the application? Because Health Server is always mandatory in all mesh applications. That should technically allow shell.c to fetch the instance and then redirect the original callback (if any defined) by using its own callback.

I think this instantiation should not be here, but rather the instance is fetched from the application using access APIs.

Andrewpini replied:
Ok, so you want the shell to overwrite the application CB's with an internal shell callback which again calls the application CB? In that case this was suggested for the prov improvments for shell, but @alxelax pointed out some risks/issues by doing so: https://github.com/zephyrproject-rtos/zephyr/pull/50180#pullrequestreview-1105635127. Another challenge is that there can be several instances of the health server as defined by the profile spec. Should the shell then only overwrite the root element health server, or should it also interfer with any secondary instances?

omkar3141 replied:
I think to get full transparency of the shell, the shell can override all callback instances. That way shell can intercept the callbacks, do its thing (print or update internal states) and pass on the info to the application. 

This will enable application fault logic testing, if any, implemented in the application and become a handy tool.

omkar3141 commented on subsys/bluetooth/mesh/shell/Kconfig:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2022 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config BT_MESH_SHELL_HEALTH_SRV
```
```suggestion
config BT_MESH_SHELL_ADD_HEALTH_SRV_INSTANCE
```

Andrewpini replied:
How about `BT_MESH_SHELL_HEALTH_SRV_INSTANCE`? I think the `ADD` is implicit in this case, and if the user is wodering they could/should always read the help box.

Andrewpini committed:
Bluetooth: Mesh: Shell: Add cfg opt for health srv

Adds Kconfig option for the Mesh Shell Health server.

The shell implementation of Health Server test commands is
dependent on the user adding and initializing the shell Health
Server instance in their application to make the commands functional.
To prevent any confussion about how these commands work, a Kconfig
option has been added to allow conditional compilation of these
commands. In this way the user must actively enable this feature, making
the conditional requirements more clear.

Signed-off-by: Anders Storrø <anders.storro@nordicsemi.no>

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -138,9 +139,14 @@ static const struct bt_mesh_health_srv_cb health_srv_cb = {
 };
+#endif /* CONFIGBT_MESH_SHELL_HEALTH_SRV_INSTANCE */
+
+
```
Nit-picking: is there a need in 2 extra empty lines?

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -1414,11 +1422,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(prov_cmds,
 #endif
 	SHELL_SUBCMD_SET_END);
 
+#if defined(CONFIGBT_MESH_SHELL_HEALTH_SRV_INSTANCE)
```
Was this checked?
```suggestion
#if defined(CONFIG_BT_MESH_SHELL_HEALTH_SRV_INSTANCE)
```

PavelVPV commented on subsys/bluetooth/mesh/shell/shell.c:
```diff
@@ -1429,8 +1439,9 @@ SHELL_STATIC_SUBCMD_SET_CREATE(test_cmds,
 #endif
 	SHELL_CMD_ARG(rpl-clear, NULL, NULL, cmd_rpl_clear, 1, 0),
-
+#if defined(CONFIGBT_MESH_SHELL_HEALTH_SRV_INSTANCE)
```
```suggestion
#if defined(CONFIG_BT_MESH_SHELL_HEALTH_SRV_INSTANCE)
```

PavelVPV commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -876,6 +876,19 @@ config BT_MESH_SHELL
+if BT_MESH_SHELL
+
+config BT_MESH_SHELL_HEALTH_SRV_INSTANCE
+	bool "Support for Shell Health Server Model"
```
```suggestion
	bool "Support for Shell Health Server model instance"
```
